### PR TITLE
EditableTree: generic access

### PR DESCRIPTION
**Pull request by sharptrip**
This PR provide a way to access fields of EditableTree without unwrapping. Currently, this works only on "all or nothing" basis meaning that (given we got an instance of EditableTreeContext with `getEditableTreeContext()`):
- one can call `context.unwrappedRoot` to get root of EditableTree with all children fields unwrapped into just values of their nodes if those are primitives and the field has a non-sequence schema, nodes with a sequence primary field ("arrays") unwrapped into arrays and so on.
- and one can call `context.root` to get root field typed as `EditableField`, where no unwrapping happens, and then iterate over its nodes, which is just an array of EditableTrees.

In turn, one can iterate over `EditableTree` to get its fields as EditableFields.

Please, check the `"traverse a complete tree (generic)"` test case and the corresponding `expectFieldEquals` and `expectNodeEquals` helper functions to see how this works.

Opinions, comments, suggestions are very welcomed.

**Step 3** in a series to push code from #12182 to enable editing of EditableTree.

[Step 2](https://github.com/microsoft/FluidFramework/pull/12306)
[Step 1](https://github.com/microsoft/FluidFramework/pull/12300)

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTreeSequence.ts [hunk 1]
```diff
@@ -0,0 +1,539 @@
+// /*!
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/editable-tree/editableTreeSequence.ts:1]: This file shouldn't be included in this PR.

File: api-report/tree.api.md [hunk 2]
```diff
@@ -233,19 +233,23 @@ export interface DetachedField extends Opaque<Brand<string, "tree.DetachedField"
...
+export type EditableField = readonly [FieldSchema, FieldKey, readonly EditableTree[]];
+
+// @public
+export interface EditableTree extends Iterable<EditableField> {
     readonly [anchorSymbol]: Anchor;
     readonly [getTypeSymbol]: (key?: FieldKey, nameOnly?: boolean) => NamedTreeSchema | TreeSchemaIdentifier | undefined;
     readonly [proxyTargetSymbol]: object;
     readonly [valueSymbol]: Value;
-    readonly [key: FieldKey]: UnwrappedEditableField;
+    readonly [key: FieldKey]: UnwrappedEditableField | EditableField;
```
Comment (CraigMacomber) [inline api-report/tree.api.md:244]: If someone calls this and gets an array back, how are they supposed to know if the array is an array of children, or an EditableField array?

I think it makes sense to leave this return type as it was, and only use EditableField for APIs where the code receiving the value didn't just specify the key (So just use it when iterating for now).

This would also kelp keep the changes in this PR focused on one thing (allowing iteration).

**Comment by CraigMacomber** [comment]
This PR seems to be doing a couple of things, and also has extra changes that were included in other PRs (which will go away when main is merged into this), as well as a file of commented out code.

Given all this, it's a bit hard to tell what the intention behind this set of changes is and the PR description doesn't help much.

EditableTree is an abstraction which hides forests from its user, so I'm not clear what "This PR enables generic access to the forest using EditableTree, see EditableField." means. I was expecting something more like one of:

- "Adds support for iterating over EditiableTree nodes for easier and faster tree enumeration"
- "Adds a way to access fields without unwrapping via getWrappedField symbol for simpler non schema aware tree traversal"
- "Replace the child arrays used in the EditableTree API with a custom type to enable them to be directly editable"

Those all say what they are doing and why, and do a single targeted change. This makes reviewing them easier. If it's clear why a change is being made, I won't accidently make suggestions that would make it no longer accomplish your goals and can help find simpler ways to meet your needs.

Also, all those example PRs would be mostly independent so they could be done on unrelated branches and reviewed concurrently without major conflicts or unrelated changes showing up in the diffs (our GitHub workflow is poor for concurrent PRs which depend on each other: it's better to try and factor out things that are independent for concurrent review).

**Comment by sharptrip** [comment]
> This PR seems to be doing a couple of things, and also has extra changes that were included in other PRs (which will go away when main is merged into this), as well as a file of commented out code.
> Given all this, it's a bit hard to tell what the intention behind this set of changes is and the PR description doesn't help much.

Of course, this PR should be reviewed after the changes from other PR are merged. Sorry for not removing the garbage file. I wanted to get an early feedback to not waste another 24 hours due to different time zones, didn't expect that having some leftovers might be such a blocker.

> EditableTree is an abstraction which hides forests from its user, so I'm not clear what "This PR enables generic access to the forest using EditableTree, see EditableField." means. I was expecting something more like one of:
> 
> * "Adds support for iterating over EditiableTree nodes for easier and faster tree enumeration"
> * "Adds a way to access fields without unwrapping via getWrappedField symbol for simpler non schema aware tree traversal"
> * "Replace the child arrays used in the EditableTree API with a custom type to enable them to be directly editable"
> 
> Those all say what they are doing and why, and do a single targeted change. This makes reviewing them easier. If it's clear why a change is being made, I won't accidently make suggestions that would make it no longer accomplish your goals and can help find simpler ways to meet your needs.

This PR is not only about to implement an iteration over fields. It allows you either to get a generic representation of the tree based on `EditableField` (`context.root`) or an old unwrapped representation (`context.unwrappedRoot`), so all or nothing. It also does not include replacing child arrays with a custom type.

I intentionally avoided to use something like `getWrappedField` as  it's not really solving the problem of unwrapped roots. As for the non-root nodes, it's just a matter of taste, imho, as same as with `getWrappedField` could be achieved if we implement an editing API like `parent[setValueSymbol](key, value)`, whereas to create new nodes one anyway has to use a parent field in something like `parent[insertField](field, data)`, so `getWrappedField` won't help in this case as well.

File: api-report/tree.api.md (hunk 2 above)
Comment (sharptrip) [reply api-report/tree.api.md]: Indeed, I was not sure about this, as for me it seems natural to have an iterator iterating over the same type as the field itself, but if you think it's fine, it's also fine for me, makes things simpler.

So, removed.

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts [hunk 3]
```diff
@@ -468,16 +471,9 @@ export function getEditableTreeContext(forest: IEditableForest): EditableTreeCon
...
-export function isArrayField(field: UnwrappedEditableField): field is UnwrappedEditableTree[] {
-    return Array.isArray(field);
-}
-
 /**
  * Checks the type of an UnwrappedEditableField.
  */
 export function isUnwrappedNode(field: UnwrappedEditableField): field is EditableTree {
-    return typeof field === "object" && !isArrayField(field);
+    return typeof field === "object" && field !== null && !Array.isArray(field);
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:478]: I don't think this change is needed: you didn't add any use of null did you?

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts [hunk 4]
```diff
@@ -83,7 +89,34 @@ export class ProxyContext implements EditableTreeContext {
...
+        return proxifyField(
+            rootSchema,
+            symbolFromKey(rootFieldKey),
+            targets,
+            true,
+        ) as UnwrappedEditableField;
+    }
+
+    // This code is kept independent of `get unwrappedRoot()` (i.e. duplicated) on purpose,
+    // as it will probably differ a lot in editing case.
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts:101]: Why not just have a private method that gets the root and takes a Boolean for unwrap or not, and have both this and unwrappedRoot call it?

Part of the motivation for making roots work the way they do is editing them should not be significantly different from editing any other field (the have a field schema, and children just like other fields), so I don't think duplication here should be necessary. Once we have proper types for fields that are not unwrapped, those should work fine for editing the root.

File: packages/dds/tree/src/test/feature-libraries/editable-tree/editableTree.spec.ts [hunk 5]
```diff
@@ -100,13 +99,20 @@ describe("editable-tree", () => {
...
         assert.equal(proxy.address[getTypeSymbol](brand("phones")), phonesSchema.name);
+        assert.equal(proxy.address[getTypeSymbol](brand("sequencePhones")), undefined);
     });
 
     it("traverse a complete tree", () => {
         const [schema, typedProxy] = buildTestPerson();
         expectTreeEquals(schema, typedProxy, personData);
     });
 
+    it("traverse a complete tree (generic)", () => {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/feature-libraries/editable-tree/editableTree.spec.ts:110]: I don't think "generic" really clarifies how this is different from the test above. Both work for all trees, so the normal English meaning of generic does not seem to be what matters here, and neither use the typescript generic types feature, so that doesn't seem to be it either.

The 3 tree comparing utility functions probably need better names or at least documentation to clarify what they do and how they are different.

Maybe "traverse by iteration" and "traverse by field keys" would make better test names?

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts (hunk 4 above)
Comment (sharptrip) [reply packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts:101]: As more I look at it (and also already having an implementation for sequences, where it didn't change at all), as more I agree on that it should not differ so much. Ok, I'll fix this.

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts (hunk 3 above)
Comment (sharptrip) [reply packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:478]: No, I didn't. Thought it might be better for completeness. Do you see any issues here (maybe, semantically)?

File: packages/dds/tree/src/test/feature-libraries/editable-tree/utils.ts [hunk 6]
```diff
@@ -56,21 +59,31 @@ export function expectTreeEquals(
...
-        const fields = expected.fields ?? {};
-        assert.equal(key in fields, true);
-        const field: JsonableTree[] = fields[key];
+    for (const ok of Reflect.ownKeys(node)) {
+        const key: FieldKey = brand(ok);
+        const subNode = node[key];
+        const expectedFields = expected.fields ?? {};
+        const expectedGlobalFields = expected.globalFields ?? {};
+        let expectedField: JsonableTree[];
+        if (isGlobalFieldKey(key)) {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/feature-libraries/editable-tree/utils.ts:68]: Use getGenericTreeField instead of handling global and local keys differently.

File: packages/dds/tree/src/test/feature-libraries/editable-tree/utils.ts [hunk 7]
```diff
@@ -56,21 +59,31 @@ export function expectTreeEquals(
...
     const type = node[getTypeSymbol](undefined, false);
     assert.deepEqual(type, expectedType);
-    for (const key of Object.keys(node)) {
-        const subNode: UnwrappedEditableField = node[key as FieldKey];
-        assert(subNode !== undefined, key);
-        const fields = expected.fields ?? {};
-        assert.equal(key in fields, true);
-        const field: JsonableTree[] = fields[key];
+    for (const ok of Reflect.ownKeys(node)) {
+        const key: FieldKey = brand(ok);
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/feature-libraries/editable-tree/utils.ts:63]: I think this comparison does not detect fields that are in the jsonableTree but missing from EditiableTrees. We can fix this by removing each field as we see it from an expectedField array:

```suggestion
    const expectedFields = new Set(genericTreeKeys(node));
    for (const ok of Reflect.ownKeys(node)) {
        const key: FieldKey = brand(ok);
        assert(expectedFields.delete(key);
```

then assert the set is empty after the loop ends.

Comment (sharptrip) [reply packages/dds/tree/src/test/feature-libraries/editable-tree/utils.ts:70]: Thanks, have missed that. Fixed.

File: packages/dds/tree/src/test/feature-libraries/editable-tree/editableTree.spec.ts (hunk 5 above)
Comment (sharptrip) [reply packages/dds/tree/src/test/feature-libraries/editable-tree/editableTree.spec.ts]: Done.

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts [hunk 8]
```diff
@@ -468,16 +471,9 @@ export function getEditableTreeContext(forest: IEditableForest): EditableTreeCon
     return new ProxyContext(forest);
 }
 
-/**
- * Checks the type of an UnwrappedEditableField.
- */
-export function isArrayField(field: UnwrappedEditableField): field is UnwrappedEditableTree[] {
```
[the remark this answers is not in the document]
Comment (sharptrip) [reply packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:474]: I have put it back. I think it's also important then to have this function in tests (without exporting to tree API), which might be fine if anyway the sequence implementation will come very soon.

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts]: `null` is not permitted by the `UnwrappedEditableField` type, so it would be invalid for a caller to pass in null here. Typescript treats null and undefined separately: handling null here is not needed.

If for some reason this function had a history of invalidly getting called with null and we wanted to help with that, an assert !null could be useful to help find such bugs and fix them, but I have no reason to think people will be violating the type system and calling this function with null.

Explicitly checking for, and handling (without throwing an error) malformed data that the type system forbids should be avoided: it just adds unnecessary code (with a download size and execution cost) and makes any bugs that pass such bad values less obviously bugs.

**Review by CraigMacomber — APPROVED** [review]
